### PR TITLE
mon: don't blow away bootstrap-mgr on upgrades

### PR DESCRIPTION
**Pull request by jcsp**
Fixes: http://tracker.ceph.com/issues/20950
Signed-off-by: John Spray <john.spray@redhat.com>

### CODE REVIEW

**Comment by jcsp** [comment]
@vasukulkarni please could you try the upgrade suite on this?

File: src/mon/AuthMonitor.cc [hunk 1]
```diff
@@ -1551,8 +1551,12 @@ void AuthMonitor::upgrade_format()
...
-    // add bootstrap key
-    {
+    // add bootstrap key if it does not already exist
+    // (might have already been get-or-create'd by
+    //  ceph-create-keys)
+    EntityName bootstrap_mgr_name;
+    bootstrap_mgr_name.from_str("client.bootstrap-mgr");
+    if (!mon->key_server.contains(bootstrap_mgr_name)) {
       KeyServerData::Incremental auth_inc;
       bool r = auth_inc.name.from_str("client.bootstrap-mgr");
```
Comment (jecluis) [inline src/mon/AuthMonitor.cc]: can't we reuse `bootstrap_mgr_name` here for `auth_inc.name` ?

Comment (jcsp) [reply src/mon/AuthMonitor.cc]: Sure!  done.

**Comment by vasukulkarni** [comment]
@jcsp sure, I will remove the workaround I have added here (https://github.com/ceph/ceph/blob/master/qa/tasks/ceph_deploy.py#L767-L779 ) and retest with that branch and update the results here.

**Comment by vasukulkarni** [comment]
@jcsp will need the shaman build , can you please push this to cephci please

**Comment by vasukulkarni** [comment]
Running here: http://pulpito.front.sepia.ceph.com/vasu-2017-10-19_20:08:39-upgrade:jewel-x:ceph-deploy:-wip-20950-distro-basic-ovh/

**Comment by vasukulkarni** [comment]
Haven't seen this before, after upgrade, it tries to restart service but it failed, after the service is restarted the mgr node is installed, I can try to shuffle the restart but it looks new since it worked before.

http://qa-proxy.ceph.com/teuthology/vasu-2017-10-19_20:39:33-upgrade:jewel-x:ceph-deploy:-wip-20950-distro-basic-vps/1752183/teuthology.log

```
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ]   python-cephfs.x86_64 2:13.0.0-2169.g57229ea.el7
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ]   python-rados.x86_64 2:13.0.0-2169.g57229ea.el7
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ]   python-rbd.x86_64 2:13.0.0-2169.g57229ea.el7
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ]
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ] Replaced:
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ]   libcephfs1.x86_64 1:10.2.10-0.el7
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ]
2017-10-19T21:12:26.248 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ] Complete!
2017-10-19T21:12:26.367 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][INFO  ] Running command: sudo ceph --version
2017-10-19T21:12:26.492 INFO:teuthology.orchestra.run.vpm011.stderr:[vpm181][DEBUG ] ceph version 13.0.0-2169-g57229ea (57229ea2a4369518c7a16b7a09b045b7896f5a70) mimic (dev)
2017-10-19T21:12:26.498 INFO:teuthology.orchestra.run.vpm181:Running: 'sudo systemctl restart ceph.target'
2017-10-19T21:12:26.594 INFO:teuthology.orchestra.run.vpm011:Running: 'sudo ceph -s'
2017-10-19T21:17:26.753 INFO:teuthology.orchestra.run.vpm011.stderr:2017-10-19 21:17:26.754 7ff3696b4700  0 monclient(hunting): authenticate timed out after 300
2017-10-19T21:17:26.790 INFO:teuthology.orchestra.run.vpm011.stderr:2017-10-19 21:17:26.754 7ff3696b4700  0 librados: client.admin authentication error (110) Connection timed out
2017-10-19T21:17:26.791 INFO:teuthology.orchestra.run.vpm011.stderr:[errno 110] error connecting to the cluster
2017-10-19T21:17:26.791 ERROR:teuthology.run_tasks:Saw exception from tasks.
```

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-10-20_15:10:22-rados-wip-kefu-testing-2017-10-20-1713-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-10-21_00:57:41-rados-wip-kefu-testing-2017-10-20-1713-distro-basic-mira/

**Comment by vasukulkarni** [comment]
tried one more run with some minor modifications but can't get it to working , maybe this is a 13.x issue, previously the upgrade was tested from jewel -> 12.x , I dont see any crash in monitors in any logs.

http://pulpito.ceph.com/vasu-2017-10-26_21:09:34-upgrade:jewel-x:ceph-deploy:-wip-20950-distro-basic-vps/

**Comment by jcsp** [comment]
@vasukulkarni can you try testing this cherry-picked onto luminous please?

**Comment by vasukulkarni** [comment]
I will pick this up on luminous and test and update here.

**Comment by vasukulkarni** [comment]
Tested this after cherry-picking the commit on top of luminous and without the mgr workaround and it is working fine,  full logs here: http://pulpito.ceph.com/vasu-2017-11-02_00:30:23-upgrade:jewel-x:ceph-deploy:-wip-qa-mgr-testing-distro-basic-vps/1800269/

**Pull request merged**